### ISSUE TITLE
IncrementalClean can delete a needed file from output

### ISSUE THREAD

**Issue by KirillOsenkov** [dotnet/msbuild#9709]
If a project has its output directory outside the project directory, and it copies a dependent file to output for more than one reason (say, via None as well as via RAR), and you remove one reason, building the project deletes the file from output, even though the file still needs to be there. Building the project again brings the file back.

1. unzip this into `C:\temp\IncrementalClean` or any other empty directory: [IncrementalClean.zip](https://github.com/dotnet/msbuild/files/14149528/IncrementalClean.zip)
2. `msbuild /r`
3. ensure `C:\temp\IncrementalClean\bin\B\A.dll` exists (hereafter it's just called "The File")
4. open `C:\temp\IncrementalClean\B\B.csproj` and comment out line 13 (to ensure the None item is not added), save file
5. `msbuild`
6. observe that The File was deleted by the `IncrementalBuild` target
7. `msbuild`
8. The File is back again

This is because `@(ReferenceCopyLocalPaths)` is not added to `FileWrites`, it is added to `FileWritesShareable`. If a .dll is copied to output because it was in `@(ReferenceCopyLocalPaths)`, it is not added to `B.csproj.FileListAbsolute.txt`.

In our repro, The File was added to FileListAbsolute.txt during step 2, because the `None` items added it to `@(FileWrites)`. In step 5, the file is copied by `_CopyFilesMarkedCopyLocal`, but since it's no longer in `@(FileWrites)`, but it is in `@(_CleanPriorFileWrites)`, `IncrementalClean` deletes it. In step 8 The File is copied back by `_CopyFilesMarkedCopyLocal`, but since it's neither in current nor prior file writes, IncrementalClean doesn't delete it.

Basically I'd say it's an unfortunate confluence of the output directory being outside the project cone and `FileWritesShareable` not being added to `FileWrites`. I think this is the problematic logic that assumes the bin folder is inside the project directory:
https://github.com/dotnet/msbuild/blob/07fd5d51f25134ea3ab3620c66f6501a74df2921/src/Tasks/Microsoft.Common.CurrentVersion.targets#L5615-L5627

I doubt there's a safe fix that we can do to fix the bug but I'm filing it just the same.

**Comment by jzabroski** [comment]
What is the recommended workaround? / @KirillOsenkov How do you workaround this bug? I think I am hitting this in an indirect way due to how Microsoft.Data.SqlClient.SNI imports native files in a net462 context. Technically, this is the guidance recommended by Microsoft. See https://learn.microsoft.com/en-us/nuget/create-packages/native-files-in-net-packages#projects-targeting-net-framework

I think the generalization is you don't need "If a project has its output directory outside the project directory" to reproduce this problem. Any clean checkout that transitively imports targets from a nuget package under `buildTransitive\<any netfx tfm>\<PackageName>.targets` will have this problem, I think.

I've tested:

* indirect references where project A ships A.exe and contains project B reference shipping B.dll _and_ a reference to Microsoft.Data.SqlClient.SNI
* direct references where project A ships A.exe and contains direct reference to Microsoft.Data.SqlClient.SNI

**Comment by KirillOsenkov** [comment]
I just add `<Target Name="IncrementalClean" />` to my Directory.Build.targets to neuter the in the box one.

If I need to clean I just git clean -xdf

**Comment by KirillOsenkov** [comment]
and yes, Microsoft.Data.SqlClient.SNI is a poorly authored NuGet package, might consider setting ExcludeAssets="all" PrivateAssets="all" GeneratePathProperty="true" and manually consuming the parts you need

**Comment by jairbubbles** [comment]
Hi @KirillOsenkov, I hit the exact same issue today.

I'm updating a NuGet package that has native .dlls. 

In the previous version, they were copied through a `.props` file (included in the package):
```xml
<?xml version="1.0" encoding="utf-8"?>
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ItemGroup>
    <NativeLibs Include="$(MSBuildThisFileDirectory)../native/*.dll" />
    <None Include="@(NativeLibs)">
      <Link>%(FileName)%(Extension)</Link>
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
  </ItemGroup>
</Project>
```

The new version of the package does not have the `.props` and the native .dlls are copied by the SDK because they are in the `native/` folder.

Nothing new here as it's really similar to what you decribed. What is new though is that **I wasn't able to reproduce the issue with the .NET 10 SDK**.

**Comment by jairbubbles** [comment]
Fixed by https://github.com/dotnet/msbuild/pull/12554 (`@(FileWritesShareable)` is now considered for the current file writes)

**Comment by baronfel** [comment]
Glad to see that fix is resolving some other long-time edge cases. I suspected we had a lot of cases where things weren't being cleaned up as expected :)

**Comment by jairbubbles** [comment]
In fact, if you test with SDK 10.0.100 , sadly `A.dll` is stil gettting deleted. 

The new behavior is enabled only when `TrackFileWritesShareableOutsideOfProjectDirectory` is true (which is the case when using the artifacts layout).

https://github.com/dotnet/sdk/blob/1b58d7631b1ed3cd26f7fa7415c075f686ae0f1a/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.DefaultOutputPaths.targets#L48-L59

TLDR; use the new artifacts layout 😊

**Comment by baronfel** [comment]
That's a defensive choice we made for compatibility more than anything - seeing examples like yours where using the flag solved _actual problems_ helps us gather evidence for changing the defaults!

(but yes, 100% use the artifacts layout if possible :D )

**Comment by jairbubbles** [comment]
Sure I understand but doesn't it seem less dangerous to not delete a file (new behavior) rather than deleting it (old behavior) 🤔?

**Comment by jzabroski** [comment]
> Sure I understand but doesn't it seem less dangerous to not delete a file (new behavior) rather than deleting it (old behavior) 🤔?

While I hate IncrementalClean (it does not play well with nuget libraries that import their own targets, due to the internal evaluation order of MSBuild requiring certain expectations of packages and their layout), not deleting a file could cause changes such as globbing things into a deployment directory that are not expected... so, it could be dangerous.

I think the artifacts layout is a pretty sane compromise, although the problem I run into is then working around limitations with the artifacts layout and any legacy CI/CD scripts I have that aren't artifacts layout-aware. You would hope in 5 years that probably just gradually rounds to 0 due to technology decay and advancement.

**Comment by baronfel** [comment]
We can also consider making this the default for SDK-based apps in .NET 11 and announcing it as a breaking change, now that it's been out in 10 and seems to be working well + solving other related problems.

@rainersigwald we should talk about the risk of this.

**Comment by jzabroski** [comment]
@baronfel , To Clarify, "making this the default" means `<TrackFileWritesShareableOutsideOfProjectDirectory>true</TrackFileWritesShareableOutsideOfProjectDirectory>` ? Or, making this work regardless of whether you use artifacts layout path?

I would personally just prefer it work this way everywhere. This one issue makes me hate IncrementalClean as a feature. It literally delayed us shifting from System.Data.SqlClient to Microsoft.Data.SqlClient by 18 months, as we did not want bespoke customizations to our process just to support buildTransitive Package.targets that didn't play well with IncrementalClean. We just decided to wait until we killed net48, as an example.

**Comment by baronfel** [comment]
@jzabroski yes - making the shareable-cleans enabled by default for all projects that use/derive from the 'base' .NET MSBuild SDK. 

I took a look at your comment [earlier in the issue](https://github.com/dotnet/msbuild/issues/9709#issuecomment-2493888366) and I'm not quite connecting the dots to how this impacted the `buildTransitive` targets in the packages you were consuming. Could you help me understand the behavior that you saw a bit more? If this feature corrects a common packaging problem that I can point to, that would make it easier for me to justify/advocate for taking the technically-breaking change.

**Comment by jzabroski** [comment]
Hi. Yes, the way that Microsoft.Data.SqlClient is packaged caused the second build operation to delete the native dlls (Microsoft.Data.SqlClient.SNI). I never fully appreciated the right fix, but I think it has to do with the order in which MSBuild evaluates projects. If you go to https://nuget.info/packages/Microsoft.Data.SqlClient.SNI/6.0.2 and open buildTransitive/net462, you will see Microsoft.Data.SqlClient.SNI.targets has a lot of interesting customizations. However, the main operation that interferes with IncrementalClean is lines 47-81, which "CopySNIFiles" throws stuff in the output directory that IncrementalClean doesn't know about.

```xml
 <!--
  ******************************************************************************
  **                          Library Content Items                           **
  ******************************************************************************
  -->

  <ItemGroup Condition="'$(ContentSNIFiles)' != '' And
                        '$(ContentSNIFiles)' != 'false' And
                        '@(SNIFiles)' != ''">
    <Content Include="@(SNIFiles)">
      <Link>%(RecursiveDir)%(FileName)%(Extension)</Link>
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </Content>
  </ItemGroup>

  <!--
  ******************************************************************************
  **                          Library Build Targets                           **
  ******************************************************************************
  -->

  <Target Name="CopySNIFiles"
          Condition="'$(CopySNIFiles)' != 'false' And
                     '$(OutDir)' != '' And
                     HasTrailingSlash('$(OutDir)') And
                     Exists('$(OutDir)')"
          Inputs="@(SNIFiles)"
          Outputs="@(SNIFiles -> '$(OutDir)%(RecursiveDir)%(Filename)%(Extension)')">
    <!--
        NOTE: Copy "Microsoft.Data.SqlClient.SNI.dll" and all related files, for every
              architecture that we support, to the build output directory.        
    -->
    <Copy SourceFiles="@(SNIFiles)"
          DestinationFiles="@(SNIFiles -> '$(OutDir)%(RecursiveDir)%(Filename)%(Extension)')" />
  </Target>

```

**Comment by jzabroski** [comment]
To elaborate what I mean by "I never fully appreciated the right fix", the [MSBuild Evaluation Order documentation ](https://learn.microsoft.com/en-us/visualstudio/msbuild/build-process-overview?view=visualstudio#:~:text=to%20the%20properties.-,Evaluation%20phase,Evaluate%20targets) leaves it unspecified exactly how the evaluation should work when the import is auto-included through a transitively referenced nuget package's buildTransitive targets.

If you look at what ultimately actually causes CopySNIFiles target to run, it is _CopySNIFilesToPublishFolder, which hooks on to _CopyFilesToPublishFolder.

```xml
  <Target Name="_CopySNIFilesToPublishFolder" AfterTargets="_CopyFilesToPublishFolder" 
            Condition="'$(CopySNIFiles)' != 'false' And
                      '$(_CopySNIFilesToPublishFolder)' != 'false' And
                      '$(_DeploymentApplicationDir)' != '' And
                      HasTrailingSlash('$(_DeploymentApplicationDir)') And
                      Exists('$(_DeploymentApplicationDir)')"
            Inputs="@(SNIDllFiles)"
            Outputs="@(SNIDllFiles -> '$(_DeploymentApplicationDir)%(RecursiveDir)%(Filename)%(Extension)')">
    <!--
        NOTE: Copy "Microsoft.Data.SqlClient.SNI.dll" only to the publish output directory.
    -->
    <Copy SourceFiles="@(SNIDllFiles)"
          DestinationFiles="@(SNIDllFiles -> '$(_DeploymentApplicationDir)%(RecursiveDir)%(Filename)%(Extension)')" />
  </Target>
```

However, this only works on the first build. It does not work on incremental builds because IncrementalClean torches the output since it is not in the known file set. 

I personally think this is not great, as it commingles data with code in both directions. If you look at how Bazel works, there's a one way data flow that builds a incremental DAG (Bazel maintains an internal, robust dependency graph of all build rules, source files, and generated outputs. It tracks every input for every build action.). That is the right architecture in my mind, but I could be wrong about it and apologize if this sounds like a dilletante view, as I am not a MSBuild expert and know you guys put a TON of effort to make my experience as a developer pain-free.
